conf/ports.py

```diff
@@ -0,0 +1,224 @@
+                else:
+                    print('Failed to create AF_XDP socket for {}. Exiting...'.format(name))
+                    sys.exit()
+
+        if conf_mode == 'af_packet':
+            try:
+                # Initialize kernel fastpath
+                kwargs = {"vdev" : "net_af_packet{},iface={},qpairs={}".format(idx, name, conf_workers), "num_out_q": conf_workers}
+                self.init_fastpath(**kwargs)
+            except:
+                print('Failed to create AF_PACKET socket for {}. Exiting...'.format(name))
+                sys.exit()
+
+        elif conf_mode == 'dpdk':
+            kwargs = None
+            pci = alias_by_interface(name)
+            if pci is not None:
+                kwargs = {"pci": pci, "num_out_q": conf_workers}
+                try:
+                    self.init_fastpath(**kwargs)
+                except:
+                    kwargs = None
+                    print('Unable to initialize {} fastpath using alias {},\
+                        falling back to scan'.format(name, pci))
+            if kwargs is None:
+                # Fallback to scanning ports
+                # if port list is empty, scan for dpdk_ports first
+                if not dpdk_ports and scan_dpdk_ports() == False:
+                    print('Registered dpdk ports do not exist.')
+                    sys.exit()
+                # Initialize DPDK fastpath
+                fidx = dpdk_ports.get(mac_by_interface(name))
+                if fidx is None:
+                    raise Exception(
+                        'Registered port for {} not detected!'.format(name))
+                kwargs = {"port_id": fidx, "num_out_q": conf_workers}
+                self.init_fastpath(**kwargs)
+
+            # Initialize kernel slowpath port and RX/TX modules
+            try:
+                peer = peer_by_interface(name)
+                vdev = "net_af_packet{},iface={}".format(idx, peer)
+                slow = PMDPort(name="{}Slow".format(name), vdev=vdev)
+                spi = __bess_module__("{}SlowPI".format(name), 'PortInc', port=slow.name)
+                spo = __bess_module__("{}SlowPO".format(name), 'PortOut', port=slow.name)
+                qspo = __bess_module__("{}QSlowPO".format(name), 'Queue')
+
+                # host_ip_filter: tcpdump -i foo 'dst host 198.19.0.1 or 198.18.0.1' -d
+                # Should always be set to lowest priority
+                HostGate = MAX_GATES - 1
+                ips = ips_by_interface(name)
+                host_ip_filter = {"priority": -HostGate, "filter": "dst host "
+                                + " or ".join(str(x) for x in ips), "gate": HostGate}
+
+                self.bpf.add(filters=[host_ip_filter])
+
+                # Direct control traffic from DPDK to kernel
+                link_modules("{}FastBPF".format(name), "{}QSlowPO".format(name), HostGate)
+                link_modules("{}QSlowPO".format(name), "{}SlowPO".format(name))
```

Comment:
can we use below as shown in https://github.com/NetSys/bess/blob/master/bessctl/sugar.py#L66-L78
Dont think retries would be necessary during initialization
```
a.connect(next_mod=b,  ogate=x)
a.connect(next_mod=b,  igate=y)
a.connect(next_mod=b,  ogate=3, igate=4)
```